File: java/breadth-first-search/OpenTheLock.java
import java.util.*;

class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> seen = new HashSet<>();

        Map<Character, List<Character>> turnMap = new HashMap<>() {
            {
                put('0', List.of('1', '9'));
                put('1', List.of('2', '0'));
                put('2', List.of('3', '1'));
                put('3', List.of('4', '2'));
                put('4', List.of('5', '3'));
                put('5', List.of('6', '4'));
                put('6', List.of('7', '5'));
                put('7', List.of('8', '6'));
                put('8', List.of('9', '7'));
                put('9', List.of('0', '8'));
            }
        };

        for(String deadend : deadends) {
            seen.add(deadend);
        }

        if(seen.contains("0000")) {
            return -1;
        }

        Queue<String> queue = new LinkedList<>();
        queue.add("0000");
        seen.add("0000");

        int turns = 0;
        while(!queue.isEmpty()) {
            int size = queue.size();
            for(int i = 0; i < size; i++) {
                String currState = queue.poll();
                if(currState.equals(target)) {
                    return turns;
                }
                
                char[] currStateArr = currState.toCharArray();
                for(int j = 0; j < currStateArr.length; j++) {
                    char currNumber = currStateArr[j];
                    for(char ch : turnMap.get(currNumber)) {
                        currStateArr[j] = ch;
                        String newState = new String(currStateArr);
                        if(!seen.contains(newState)) {
                            queue.add(newState);
                            seen.add(newState);
                        }
                        currStateArr[j] = currNumber;
                    }
                }
            }
            turns += 1;
        }
        return -1;
    }
}

File: python/breadth-first-search/OpenTheLock.py
from typing import List
from collections import deque

class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        seen = set(deadends)

        turn_map = {
            '0': ['1', '9'],
            '1': ['2', '0'],
            '2': ['3', '1'],
            '3': ['4', '2'],
            '4': ['5', '3'],
            '5': ['6', '4'],
            '6': ['7', '5'],
            '7': ['8', '6'],
            '8': ['9', '7'],
            '9': ['0', '8']
        }

        if "0000" in seen:
            return -1

        queue = deque(["0000"])
        seen.add("0000")

        turns = 0
        while queue:
            size = len(queue)
            for _ in range(size):
                curr_state = queue.popleft()

                if curr_state == target:
                    return turns

                for j in range(len(curr_state)):
                    curr_number = curr_state[j]
                    for ch in turn_map[curr_number]:
                        new_state = curr_state[:j] + ch + curr_state[j+1:]
                        if new_state not in seen:
                            queue.append(new_state)
                            seen.add(new_state)

            turns += 1
            
        return -1
